components/external-services-mock/internal/oauth/token_handler.go

```diff
@@ -30,133 +55,189 @@ func NewHandler(expectedSecret, expectedID string) *handler {
-		tenantHeaderName: tenantHeaderName,
-		signingKey:       signingKey,
+		expectedSecret:      expectedSecret,
+		expectedID:          expectedID,
+		tenantHeaderName:    tenantHeaderName,
+		expectedUsername:    expectedUsername,
+		expectedPassword:    expectedPassword,
+		signingKey:          signingKey,
+		staticMappingClaims: staticMappingClaims,
 	}
 }
 
 func (h *handler) Generate(writer http.ResponseWriter, r *http.Request) {
-	authorization := r.Header.Get("authorization")
-	id, secret, err := getBasicCredentials(authorization)
-	if err != nil {
-		log.C(r.Context()).Errorf("client secret not found in header: %s", err.Error())
-		httphelpers.WriteError(writer, errors.Wrap(err, "client secret not found in header"), http.StatusBadRequest)
+	ctx := r.Context()
+
+	if r.Header.Get(ContentTypeHeader) != ContentTypeApplicationURLEncoded {
+		log.C(ctx).Errorf("Unsupported media type, expected: application/x-www-form-urlencoded got: %s", r.Header.Get(ContentTypeHeader))
+		writer.WriteHeader(http.StatusUnsupportedMediaType)
 		return
 	}
 
-	if h.expectedID != id || h.expectedSecret != secret {
-		log.C(r.Context()).Error("client secret or client id doesn't match expected")
-		httphelpers.WriteError(writer, errors.New("client secret or client id doesn't match expected"), http.StatusBadRequest)
+	if err := r.ParseForm(); err != nil {
+		log.C(ctx).WithError(err).Error("An error occurred while parsing query")
+		httphelpers.WriteError(writer, errors.New("An error occurred while parsing query"), http.StatusInternalServerError)
 		return
 	}
-	h.GenerateWithoutCredentials(writer, r)
+
+	if r.FormValue(GrantTypeFieldName) != CredentialsGrantType && r.FormValue(GrantTypeFieldName) != PasswordGrantType {
+		log.C(ctx).Errorf("The grant_type should be %s or %s but we got: %s", CredentialsGrantType, PasswordGrantType, r.FormValue(GrantTypeFieldName))
+		httphelpers.WriteError(writer, errors.New("An error occurred while parsing query"), http.StatusBadRequest)
+		return
+	}
+
+	if r.FormValue(GrantTypeFieldName) == CredentialsGrantType {
+		if err := h.handleClientCredentialsRequest(r); err != nil {
+			log.C(ctx).Error(err)
+			httphelpers.WriteError(writer, err, http.StatusBadRequest)
+			return
+		}
+	} else { // Assume it's a password flow because currently we support only client_credentials and password
+		if err := h.handlePasswordCredentialsRequest(r); err != nil {
+			log.C(ctx).Error(err)
+			httphelpers.WriteError(writer, err, http.StatusBadRequest)
+			return
+		}
+	}
+
+	claims := make(map[string]interface{})
+	claimsFunc, ok := h.staticMappingClaims[r.FormValue(ClaimsKey)]
+	if ok { // If the request contains claims key, use the corresponding claims in the static mapping for that key
+		claims = claimsFunc()
+	} else { // If there is no claims key provided use empty claims
+		log.C(ctx).Info("Did not find claims key in the request. Proceeding with empty claims...")
+	}
+
+	claims[ZidKey] = r.Header.Get(h.tenantHeaderName)
```

Comment:
There is no specific reason, that was the way it was handled before. But I'm not sure how it can be part of the claimsFunc because we get the zid claim from request header and claimsFunc is declared in main.go of the ext-svc-mock